[FIX #3074] Combine PR scripts

sebastiandelgado opened the pull request:
Fixes #3074 

### Summary:

- Combine merge-pr.sh and merge-external-pr.sh using merge-external-pr.sh as a base.
- Modify script interface to `./scripts/merge-pr.sh <PR-ID>` and remove options to run with `<REPO_URL> <BRANCH>` arguments.
- Works regardless of where PR is hosted (internal/external) and whether or not it is signed.
- Keeps core team signature requirement.

Status: Complete

yenda commented on scripts/merge-pr.sh:
```diff
@@ -15,30 +20,127 @@ confirm() {
+  git rebase $BRANCH
+}
+
+check_is_pr_single_commit() {
```
so here the idea is that you need to confirm that you want to merge if it is not a single commit because most of the time we want to squash the commits in a PR before merging, but it might be the case that we want separate commits, so there need to be a confirmation but the script should not just fail on it.

yenda replied:
To test: 
if you have 2 commits in a PR, script should ask for confirm
if you have 1 commit, it should just do the merge

sebastiandelgado replied:
Got it. Will update soon.

sebastiandelgado commented:
Just tested and added a commit that adds the option for merging multiple commits instead of rejecting anything that isn't squashed. Will squash this PR when its ready.

oskarth commented:
Thanks @sebastiandelgado!

>  Status: WIP. Waiting for clarification on several issues over at #3074

Could you write the outstanding questions here? Seems like most questions got resolved.

yenda commented:
@sebastiandelgado Is it ready or is there still something you are working on ? Please squash the commits if it is ready. Optionnaly you can sign your commit if you know how to use gpg signatures but it is not a requirement for external contributors since a core contributor will sign them when we merge.

Commit:
[FIX #3074] Combine PR scripts and clean up usage of merge-external-pr.sh

    - Combine merge-pr.sh and merge-external-pr.sh using merge-external-pr.sh as a base
    - Modify script interface to ./scripts/merge-pr.sh <PR-ID> and remove options to run with REPO_URL BRANCH arguments
    - Works regardless of where PR is hosted and whether or not it is signed by external contributor.

Signed-off-by: Sebastian Delgado <sdalcega@gmail.com>
Signed-off-by: Eric Dvorsak <eric@dvorsak.fr>

sebastiandelgado commented:
@oskarth @yenda its ready to be landed. I just squashed the 3 commits. Let me know if anything else needs to be done on my side.

andytudhope reviewed (approved):
LGTM

sebastiandelgado commented:
Thanks guys!

yenda commented:
successfully merged with the script of the PR

yenda commented:
Actually when trying to merge a PR of mine I get this now:

```
Writing objects: 100% (12/12), 1.92 KiB | 0 bytes/s, done.
Total 12 (delta 9), reused 0 (delta 0)
remote: Resolving deltas: 100% (9/9), completed with 9 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/develop.
remote: error: At least one approved review is required by reviewers with write access.
To git@github.com:status-im/status-react.git
 ! [remote rejected] develop -> develop (protected branch hook declined)
error: failed to push some refs to 'git@github.com:status-im/status-react.git'
```

sebastiandelgado commented:
@yenda what branch are you trying to merge so I can try to replicate the conditions? Have you tried using the original merge-pr.sh (not the merge-external-pr.sh or the new merge-pr.sh)?

yenda commented:
@sebastiandelgado I tried to merge https://github.com/status-im/status-react/pull/3120 with the new merge-pr script

yenda commented:
I think it has something to do with 

```
  if [[ $(echo "$pr_info" | jq -r .maintainer_can_modify) == true ]]; then
    RW_PR_REPO=1
  else
    warn "PR does not allow 'edits from maintainers', so it will be kept open"
  fi
```

yenda commented on scripts/merge-pr.sh:
```diff
@@ -15,30 +20,127 @@ confirm() {
+  load_config
+  check_pr_prereq
+  get_pr_info "$@"
+  cleanup
```
why is cleanup happening so early ?

sebastiandelgado replied:
This cleanup statement is at the same place it was at originally. The intention here is to checkout develop and remove any leftover branches created in a previous failed attempt at running this script with the same PR-id.

sebastiandelgado commented:
@yenda are you getting the 
`PR does not allow 'edits from maintainers', so it will be kept open`
warning before you get the 
`remote: error: At least one approved review is required by reviewers with write access.`
 error? 

My understanding from the original state of the script is that PRs (at least external ones) on this repo were required to enable the `maintainer_can_modify` flag from github. 

Maybe thats not the case for core contributors? #3120 has `maintainer_can_modify=false`. The point of checking for this flag is so that the external branch can be renamed to `pr-<PRid>`. So correct me if I'm wrong, but it seems that this is done for uniformity.

As far as I can tell, there are three fixes:

1. Change merge-pr.sh so that it requires that all PRs (including those by core contributors) have `maintainer_can_modify=true` by using `fatal "PR does not allow 'edits from maintainers', so it will be kept open"` instead of just using `warn` on line 59, and removing the conditional on line 104.
2. Change merge-pr.sh such that `$PR_REMOTE_NAME` and `$PR_LOCAL_BRANCH` get set to the original name of the branch (not `pr-<PRid>`), which is available at `echo "pr_info" | jq -r .head.ref`
3. Don't even check the `maintainer_can_modify` flag and always use the original branch name instead of `pr-<PRid>`

I'm more than happy to issue a second PR with whichever change you prefer. Let me know what you think!

yenda commented:
@sebastiandelgado  wouldn't it also be possible to differentiate the external and core contributors PR ?

yenda commented:
@sebastiandelgado I am getting the warning before yes sorry I missed the question.

sebastiandelgado commented:
@yenda thanks for clarifying. You're right about being able to differentiate between external and core contributors from the github API.

I made a new PR #3155 which checks for this condition. I tested it on my dummy repo which I think has the same settings as status-im/status-react (including the required review), so it *should* work. 

Maybe its worth it for you to test this when trying to merge another PR before we approve it? Let me know what you think.
